### PR TITLE
Enable completion of object initializer of generic type constraint.

### LINKED ISSUE #24612: New generic class, No prompts if initialization
The generic T constraint is a subclass of an interface with parameterless constructors, both A and B are properties of the interface new t{a=&quot;&quot;,b=&quot;&quot;}; The IDE does not hint

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/173607/new泛型类时若进行初始化则无提示.html
VSTS ticketId: 544394_
_These are the original issue comments:_

Ariel Yang - MSFT on ‎12‎/‎26‎/‎2017, 08:36 PM (37 days ago): <p>Translated from bing :</p><p>A generic t constraint is a subclass of an interface with a parameterless constructor, and A and B are all properties of an interface
<br>New T\{a= "", b= ""};
<br>The IDE does not hint<strong></strong><u></u><sub></sub><sup></sup></p>

Jinu Joseph [MSFT] on ‎1‎/‎7‎/‎2018, 06:00 PM (25 days ago): <p>We appreciate you taking the time to report this problem. We are currently prioritizing problems that are impacting a broad set of our customers, so we may not be able to investigate this one immediately. We know this problem is important to you, so we will continue to monitor it.</p>

_These are the original issue solutions:_
(no solutions)

### PR DESCRIPTION
**Pull request by zaytsev-victor**
Fixes #24612
<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectInitializerCompletionProvider.cs [hunk 1]
```diff
@@ -125,10 +125,13 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
                     return null;
                 }
 
-                var ctor = semanticModel.GetSymbolInfo(objectCreation, cancellationToken).Symbol;
-                var type = ctor != null ? ctor.ContainingType : null;
+                var type = semanticModel.GetSymbolInfo(objectCreation.Type, cancellationToken).Symbol as ITypeSymbol;
+                if (type is ITypeParameterSymbol typeParameterSymbol)
+                {
+                    return Tuple.Create<ITypeSymbol, Location>(typeParameterSymbol.GetNamedTypeSymbolConstraint(), token.GetLocation());
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectInitializerCompletionProvider.cs]: what if ```typeParameterSymbol.GetNamedTypeSymbolConstraint()``` returns null?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs [hunk 2]
```diff
@@ -820,6 +820,33 @@ static void Main(string[] args)
...
+    {
+        return new T
+        {
+            $$
+        };
+    }
+}";
+
+            await VerifyItemExistsAsync(markup, "A");
+            await VerifyItemExistsAsync(markup, "B");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: can we have negative tests, where T does not have a class/interface constraint?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs [hunk 3]
```diff
@@ -847,6 +847,26 @@ public static T Create<T>()
...
 
+        [WorkItem(24612, "https://github.com/dotnet/roslyn/issues/24612")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task ObjectInitializerOfGenericTypeСonstraint2()
+        {
+            var markup = @"
+internal class Example
+{
+    public static T Create<T>()
+        where T : new()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: can we also do "where T : struct" and "where T : System.Delegate" and "where T : unmanaged"?

Comment (zaytsev-victor) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: @CyrusNajmabadi, I'll add a test with ```where T : struct``` constraint.
Regarding other cases:
1. ```where T : System.Delegate```: "Constraint cannot be special class 'Delegate'"
2. ```where T : unmanaged```: I'm not sure what do you mean. It is not valid syntax.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: I thought it was.  See: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.3/blittable.md

@OmarTawfik can you advise?  Has this been added to the language?

Comment (OmarTawfik) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: Yes. You've to use C# 7.3 for the new syntax to work :)
Please LMK/mention me if you see any issues.

**Comment by zaytsev-victor** [comment]
retest please

**Comment by sharwell** [comment]
retest this please

**Comment by sharwell** [comment]
@jinujoseph Requesting 16.0 Preview 1

**Comment by jinujoseph** [comment]
I think we should keep this in master

**Pull request merged**